### PR TITLE
autotest: Simplify some test skipping

### PR DESCRIPTION
**Pull request by QuLogic**
## What does this PR do?

Skips due to import errors can be done using `pytest.importorskip`, which will give a standard message and be a little shorter. It's also a bit nicer to use `pytest.mark.skipif` where possible, and actually pass skip reasons to `pytest.skip` instead of printing them.

## Tasklist

 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: autotest/gcore/testnonboundtoswig.py [hunk 1]
```diff
@@ -96,12 +92,12 @@ def setup():
...
+                        f'static version ({static_version})')
 
         return gdal_handle
     except Exception:
-        print('cannot find gdal shared object')
-        pytest.skip()
+        pytest.skip('cannot find gdal shared object')
 
 ###############################################################################
 # Call GDALDestroyDriverManager()
```
Comment (rcoup) [inline autotest/gcore/testnonboundtoswig.py:103]: Seems like the cleanup below methods aren't called?

File: autotest/gcore/testnonboundtoswig.py [hunk 2]
```diff
@@ -32,14 +32,10 @@
...
-except ImportError:
-    pass
-
 import gdaltest
 
+ctypes = pytest.importorskip('ctypes')
+
 gdal_handle_init = False
 gdal_handle = None
 gdal_handle_stdcall = None
```
Comment (rcoup) [inline autotest/gcore/testnonboundtoswig.py:41]: It's really unclear why we're loading both cdll + windll on windows, then randomly alternating calls between them? 

I wonder if it's better to restructure this as a fixture that returns a different handle depending on OS, then use that in the test methods below.

File: autotest/gcore/tiff_read.py [hunk 3]
```diff
@@ -2742,12 +2737,9 @@ def test_tiff_read_huge_implied_number_strips():
 ###############################################################################
 
 
+@pytest.mark.skipif(sys.platform == 'win32', reason='Runs super slow on some Windows configs')
 def test_tiff_read_many_blocks():
 
-    # Runs super slow on some Windows configs
```
Comment (rcoup) [inline autotest/gcore/tiff_read.py:2747]: worth re-validating this? Was added 3 years ago in 2cdfe800bc6876ec3c60bb305802dc5b566bc6f8

File: autotest/gcore/virtualmem.py [hunk 4]
```diff
@@ -31,24 +31,18 @@
...
-    numpy_available = True
-except ImportError:
-    numpy_available = False
-
 from osgeo import gdal
 import pytest
 
-###############################################################################
-# Test linear and tiled virtual mem interfaces in read-only mode
+numpy = pytest.importorskip('numpy')
```
Comment (rcoup) [inline autotest/gcore/virtualmem.py:37]: `pytestmark =` isn't particularly obvious it's a module-level skip
```suggestion
# All tests will be skipped if numpy isn't available or SKIP_VIRTUALMEM is set)
numpy = pytest.importorskip('numpy')
```

File: autotest/gcore/numpy_rw.py [hunk 5]
```diff
@@ -29,19 +29,13 @@
...
-except ImportError:
-    numpy_available = False
-
 import gdaltest
 from osgeo import gdal
 import pytest
 
 
-pytestmark = pytest.mark.skipif(not numpy_available, reason='numpy not available')
+numpy = pytest.importorskip('numpy')
```
Comment (rcoup) [inline autotest/gcore/numpy_rw.py:37]: `pytestmark =` isn't particularly obvious it's a module-level skip:
```suggestion
# All tests will be skipped if numpy or gdal_array aren't available
numpy = pytest.importorskip('numpy')
```

File: autotest/gcore/virtualmem.py [hunk 6]
```diff
@@ -167,13 +153,8 @@ def test_virtualmem_3():
 
 ###############################################################################
 # Test virtual mem auto with GTiff
-
-
+@pytest.mark.skipif(not sys.platform.startswith('linux'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/virtualmem.py:156]: wonder if we should make these global helpers?

ala:
```python
skip_if_not_linux = pytest.mark.skipif(not sys.platform.startswith('linux'), reason='Not Linux')

@skip_if_not_linux
def test_1():
```

File: autotest/gcore/vsicrypt.py [hunk 7]
```diff
@@ -466,10 +466,7 @@ def test_vsicrypt_5():
 
 def test_vsicrypt_6(testnonboundtoswig_setup):  # noqa
 
-    try:
-        import ctypes
-    except ImportError:
-        pytest.skip()
+    ctypes = pytest.importorskip('ctypes')
```
Comment (rcoup) [inline autotest/gcore/vsicrypt.py:469]: when wouldn't ctypes be available?

File: autotest/gcore/vsigs.py [hunk 8]
```diff
@@ -1058,19 +1055,16 @@ def method(request):
     gdal.SetConfigOption('CPL_GCE_CREDENTIALS_URL', '')
     gdal.SetConfigOption('CPL_GCE_CHECK_LOCAL_FILES', None)
 
+
 ###############################################################################
 # Read credentials from simulated GCE instance with expiration of the
 # cached credentials
-
-
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/vsigs.py:1062]:
```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gcore/vsigs.py [hunk 9]
```diff
@@ -1150,11 +1144,9 @@ def test_vsigs_extra_1():
         pytest.skip()
 
     # if gdal.GetConfigOption('GS_SECRET_ACCESS_KEY') is None:
-    #    print('Missing GS_SECRET_ACCESS_KEY')
-    #    pytest.skip()
+    #    pytest.skip('Missing GS_SECRET_ACCESS_KEY')
     # elif gdal.GetConfigOption('GS_ACCESS_KEY_ID') is None:
-    #    print('Missing GS_ACCESS_KEY_ID')
-    #    pytest.skip()
+    #    pytest.skip('Missing GS_ACCESS_KEY_ID')
```
Comment (rcoup) [inline autotest/gcore/vsigs.py:1149]: Feels like we can probably drop these altogether?

File: autotest/gcore/vsis3.py [hunk 10]
```diff
@@ -3129,18 +3129,15 @@ def test_vsis3_read_credentials_config_file_inconsistent():
     gdal.SetConfigOption('AWS_CONFIG_FILE', '')
     gdal.Unlink('/vsimem/aws_config')
 
+
 ###############################################################################
 # Read credentials from simulated EC2 instance
-
-
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/vsis3.py:3135]:
```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gcore/vsis3.py [hunk 11]
```diff
@@ -3191,18 +3188,15 @@ def test_vsis3_read_credentials_ec2_imdsv2():
     gdal.SetConfigOption('CPL_AWS_EC2_API_ROOT_URL', '')
     gdal.SetConfigOption('CPL_AWS_AUTODETECT_EC2', None)
 
+
 ###############################################################################
 # Read credentials from simulated EC2 instance that only supports IMDSv1
-
-
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/vsis3.py:3194]:
```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gcore/vsis3.py [hunk 12]
```diff
@@ -3240,19 +3234,16 @@ def test_vsis3_read_credentials_ec2_imdsv1():
     gdal.SetConfigOption('CPL_AWS_EC2_API_ROOT_URL', '')
     gdal.SetConfigOption('CPL_AWS_AUTODETECT_EC2', None)
 
+
 ###############################################################################
 # Read credentials from simulated EC2 instance with expiration of the
 # cached credentials
-
-
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/vsis3.py:3241]:
```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gdrivers/eedai.py [hunk 13]
```diff
@@ -442,14 +442,12 @@ def test_eedai_GOOGLE_APPLICATION_CREDENTIALS():
 # Read credentials from simulated GCE instance
 
 
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gdrivers/eedai.py:445]:
```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gdrivers/jpeg.py [hunk 14]
```diff
@@ -886,18 +883,15 @@ def test_jpeg_26():
...
     gdal.Unlink('/vsimem/jpeg_26.jpg')
 
+
 ###############################################################################
 # Test reading a file that contains the 2 denial of service
 # vulnerabilities listed in
 # http://www.libjpeg-jpeg_26.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf
 
-
+@pytest.mark.skipif(sys.platform == 'win32', reason='Fails for some reason on Windows')
```
Comment (rcoup) [inline autotest/gdrivers/jpeg.py:892]: xfail?

File: autotest/gdrivers/kmlsuperoverlay.py [hunk 15]
```diff
@@ -196,10 +196,7 @@ def test_kmlsuperoverlay_4():
 
 def test_kmlsuperoverlay_5():
 
-    try:
-        from xml.etree import ElementTree
-    except ImportError:
-        pytest.skip()
+    ElementTree = pytest.importorskip('xml.etree.ElementTree')
```
Comment (rcoup) [inline autotest/gdrivers/kmlsuperoverlay.py:199]: This is in stdlib these days, so shouldn't be failing

File: autotest/gdrivers/mem.py [hunk 16]
```diff
@@ -108,10 +108,7 @@ def test_mem_2():
     gdal.PopErrorHandler()
     assert ds is None, 'opening MEM dataset should have failed.'
 
-    try:
-        import ctypes
-    except ImportError:
-        pytest.skip()
+    ctypes = pytest.importorskip('ctypes')
```
Comment (rcoup) [inline autotest/gdrivers/mem.py:111]: ctypes is in stdlib

File: autotest/ogr/ogr_shape.py [hunk 17]
```diff
@@ -2771,11 +2771,9 @@ def test_ogr_shape_67():
 # Test opening a shape datasource with files with mixed case and then REPACK
 
 
+@pytest.mark.skipif(sys.platform == 'darwin', reason='Fails on MacOSX. Not sure why.')
```
Comment (rcoup) [inline autotest/ogr/ogr_shape.py:2774]:
```suggestion
@pytest.mark.xfail(sys.platform == 'darwin', reason='Fails on MacOSX. Not sure why.')
```

File: autotest/ogr/ogr_odbc.py [hunk 18]
```diff
@@ -39,6 +39,8 @@
 import ogrtest
 import pytest
 
+pytestmark = pytest.mark.skipif(sys.platform != 'win32', reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/ogr/ogr_odbc.py:42]:
```suggestion
# Skip all tests unless we're on Windows
pytestmark = pytest.mark.skipif(sys.platform != 'win32', reason='Incorrect platform')
```

File: autotest/ogr/ogr_gpkg.py [hunk 19]
```diff
@@ -4298,12 +4298,10 @@ def abortAfterDelay2():
 # Test ST_Transform() with no record in gpkg_spatial_ref_sys and thus we
 # fallback to EPSG
 
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason='f.GetGeometryRef() returns None on the current Windows CI')
```
Comment (rcoup) [inline autotest/ogr/ogr_gpkg.py:4302]:
```suggestion
@pytest.mark.xfail(sys.platform == 'win32',
                    reason='f.GetGeometryRef() returns None on the current Windows CI')
```

File: autotest/ogr/ogr_shape.py [hunk 20]
```diff
@@ -2913,11 +2909,9 @@ def test_ogr_shape_70():
 # Test heterogeneous file permissions on .shp and .dbf.
 
 
+@pytest.mark.skipif(sys.platform.find('linux') != 0, reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/ogr/ogr_shape.py:2912]:
```suggestion
@pytest.mark.skipif(sys.platform != 'linux', reason='Incorrect platform')
```

File: autotest/pyscripts/test_gdal_calc.py [hunk 21]
```diff
@@ -42,13 +42,13 @@
 from collections import defaultdict
 
 # test that numpy is available, if not skip all tests
+np = pytest.importorskip('numpy')
+gdal_calc = pytest.importorskip('osgeo.utils.gdal_calc')
+gdal_array = pytest.importorskip('osgeo.gdal_array')
```
Comment (rcoup) [inline autotest/pyscripts/test_gdal_calc.py:47]: when wouldn't gdal_calc/gdal_array be available?

File: autotest/pyscripts/test_ogr2ogr_py.py [hunk 22]
```diff
@@ -182,10 +182,7 @@ def test_ogr2ogr_py_5():
 
 def test_ogr2ogr_py_6():
 
-    try:
-        import ogr_pg
-    except:
-        pytest.skip()
+    ogr_pg = pytest.importorskip('ogr_pg')
```
Comment (rcoup) [inline autotest/pyscripts/test_ogr2ogr_py.py:185]: when wouldn't this be available?

File: autotest/pyscripts/test_gdal_calc.py (hunk 21 above)
Comment (rcoup) [reply autotest/pyscripts/test_gdal_calc.py:47]: A: when numpy wasn't available at build-time.

File: autotest/pyscripts/test_ogr2ogr_py.py (hunk 22 above)
Comment (rcoup) [reply autotest/pyscripts/test_ogr2ogr_py.py:185]: Probably needs a `reason=` argument if it needs to stay

File: autotest/pyscripts/test_ogr2ogr_py.py [hunk 23]
```diff
@@ -215,10 +212,7 @@ def test_ogr2ogr_py_6():
 
 def test_ogr2ogr_py_7():
 
-    try:
-        import ogr_pg
-    except:
-        pytest.skip()
+    ogr_pg = pytest.importorskip('ogr_pg')
```
Comment (rcoup) [inline autotest/pyscripts/test_ogr2ogr_py.py:215]: same here

File: autotest/gcore/vsigs.py [hunk 24]
```diff
@@ -983,18 +983,15 @@ def method(request):
     gdal.SetConfigOption('GOA2_AUTH_URL_TOKEN', None)
     gdal.Unlink('/vsimem/.boto')
 
+
 ###############################################################################
 # Read credentials from simulated GCE instance
-
-
+@pytest.mark.skipif(sys.platform not in ('linux', 'linux2', 'win32'), reason='Incorrect platform')
```
Comment (rcoup) [inline autotest/gcore/vsigs.py:989]: `linuxN` was dropped from Py3.3+

```suggestion
@pytest.mark.skipif(sys.platform not in ('linux', 'win32'), reason='Incorrect platform')
```

File: autotest/gcore/testnonboundtoswig.py (hunk 2 above)
Comment (rcoup) [reply autotest/gcore/testnonboundtoswig.py:41]: Could be a separate PR though

**Review by rcoup — CHANGES_REQUESTED** [review]
Nice! A few minor tweaks

File: autotest/gcore/testnonboundtoswig.py [hunk 25]
```diff
@@ -32,14 +32,10 @@
...
-
-try:
-    import ctypes
-except ImportError:
-    pass
-
 import gdaltest
 
+ctypes = pytest.importorskip('ctypes')
+
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply autotest/gcore/testnonboundtoswig.py:38]: I don't know; I assumed it was for some weird franken-Python on a relatively unused system.

File: autotest/gcore/testnonboundtoswig.py (hunk 1 above)
Comment (QuLogic) [reply autotest/gcore/testnonboundtoswig.py:103]: There's a lot of that weirdness left.

File: autotest/gcore/testnonboundtoswig.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (rcoup) [reply autotest/gcore/testnonboundtoswig.py:38]: I looked it up, `ctypes` got added to stdlib in 2.5, I suspect the original test suite predated that 😄

File: autotest/gcore/testnonboundtoswig.py (hunk 1 above)
Comment (rcoup) [reply autotest/gcore/testnonboundtoswig.py:103]: Ya, getting to pytest at all was a massive effort, but (as you're doing) there's plenty of other cruft to tidy up

File: autotest/pyscripts/test_ogr2ogr_py.py (hunk 22 above)
Comment (QuLogic) [reply autotest/pyscripts/test_ogr2ogr_py.py:185]: Not sure that `importorskip` needs a `reason`.

File: autotest/ogr/ogr_shape.py (hunk 20 above)
Comment (QuLogic) [reply autotest/ogr/ogr_shape.py:2912]: I think you have that backwards.

File: autotest/gcore/tiff_read.py (hunk 3 above)
Comment (QuLogic) [reply autotest/gcore/tiff_read.py:2747]: We'll see how CI likes it.

Comment (QuLogic) [reply autotest/gcore/tiff_read.py:2747]: Seems to be okay remove the skip.

File: autotest/gcore/testnonboundtoswig.py [hunk 26]
```diff
@@ -96,12 +92,12 @@ def setup():
...
             gdal_handle_stdcall = None
-            pytest.skip('dynamic version(%s) does not match static version (%s)' % (dynamic_version, static_version))
+            pytest.skip(f'dynamic version({dynamic_version}) does not match '
+                        f'static version ({static_version})')
 
         return gdal_handle
     except Exception:
-        print('cannot find gdal shared object')
-        pytest.skip()
+        pytest.skip('cannot find gdal shared object')
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply autotest/gcore/testnonboundtoswig.py:95]: I leave it to you all to decide on these.

**Comment by rcoup** [comment]
I'm happy with this.

Any thoughts on the above @rouault?

**Comment by rouault** [comment]
> Any thoughts on the above @rouault?

on a particular point ? or on this PR ? The PR LGTM

**Comment by rcoup** [comment]
Ah sorry, mostly whether we should be working towards using `xfail` instead of `skip` for tests which fail on a particular OS without a good reason.

**Comment by rouault** [comment]
> Ah sorry, mostly whether we should be working towards using `xfail` instead of `skip` for tests which fail on a particular OS without a good reason.

I don't have a strong opinion on this. xfail might be better if it is a hint to grep for it and investigate further.

**Comment by rouault** [comment]
@QuLogic Do you intend to do more work in this PR or maybe in follow-up PRs ?

**Comment by QuLogic** [comment]
No, nothing more, unless you have some strong opinions on the `xfail`s.

**Pull request merged**